### ISSUE TITLE
Implement ISBN lookup for Relaton

### ISSUE THREAD

**Issue by ronaldtse** [relaton/relaton-isbn#1]
OpenLibrary seems to be the best choice:

https://www.vinzius.com/post/free-and-paid-api-isbn/

e.g.
https://openlibrary.org/books/OL2843974M/The_lambda_calculus

**Comment by ronaldtse** [comment]
OpenLibrary API: https://openlibrary.org/developers/api

**Comment by ronaldtse** [comment]
https://github.com/jayfajardo/openlibrary

**Comment by ronaldtse** [comment]
OpenLibrary works!!

e.g.
https://openlibrary.org/api/books?bibkeys=ISBN:0201758393&jscmd=details&format=json

```json
{"ISBN:0201758393": {"bib_key": "ISBN:0201758393", "info_url": "https://openlibrary.org/books/OL21485047M/PDF_reference", "preview": "full", "preview_url": "https://archive.org/details/pdfreferencevers00inca", "thumbnail_url": "https://covers.openlibrary.org/b/id/8704885-S.jpg", "details": {"publishers": ["Addison-Wesley"], "identifiers": {"goodreads": ["788687"], "librarything": ["769832"]}, "subtitle": "Adobe portable document format version 1.4", "isbn_10": ["0201758393"], "covers": [8704885], "lc_classifications": ["QA76.76.T49 P38 2001", "QA76.76.T49P38 2001"], "key": "/books/OL21485047M", "ocaid": "pdfreferencevers00inca", "publish_places": ["Boston"], "contributions": ["Adobe Systems."], "languages": [{"key": "/languages/eng"}], "pagination": "xx, 945 p. :", "source_records": ["marc:marc_oregon_summit_records/catalog_files/ohsu_ncnm_wscc_bibs.mrc:105547460:1004", "ia:pdfreferencevers00inca", "bwb:9780201758399", "marc:marc_loc_2016/BooksAll.2016.part28.utf8:230431192:779"], "title": "PDF reference", "dewey_decimal_class": ["005.7/2"], "notes": {"type": "/type/text", "value": "Includes bibliographical references and index."}, "number_of_pages": 945, "edition_name": "3rd ed.", "lccn": ["2001053899"], "subjects": ["Adobe Acrobat", "Portable document software"], "publish_date": "2001", "publish_country": "mau", "by_statement": "Adobe Systems Incorporated.", "oclc_numbers": ["48144467"], "works": [{"key": "/works/OL18476752W"}], "type": {"key": "/type/edition"}, "latest_revision": 9, "revision": 9, "created": {"type": "/type/datetime", "value": "2008-11-02T19:29:56.844069"}, "last_modified": {"type": "/type/datetime", "value": "2022-11-17T13:03:45.372349"}}}}
```

**Comment by ronaldtse** [comment]
The is the API we should use:

https://openlibrary.org/dev/docs/api/read

https://openlibrary.org/api/volumes/brief/isbn/9780120644810.json

```
{
  "records": {
    "/books/OL21119585M": {
      "isbns": [
        "0120644819"
      ],
      "issns": [

      ],
      "lccns": [

      ],
      "oclcs": [

      ],
      "olids": [
        "OL21119585M"
      ],
      "publishDates": [
        "1991"
      ],
      "recordURL": "https://openlibrary.org/books/OL21119585M/Graphics_gems_II",
      "data": {
        "url": "https://openlibrary.org/books/OL21119585M/Graphics_gems_II",
        "key": "/books/OL21119585M",
        "title": "Graphics gems II",
        "authors": [
          {
            "url": "https://openlibrary.org/authors/OL2646519A/James_Arvo",
            "name": "James Arvo"
          }
        ],
        "number_of_pages": 643,
        "pagination": "xxxii,643p.,(4)p. of plates :",
        "by_statement": "edited by James Arvo.",
        "identifiers": {
          "goodreads": [
            "3994749"
          ],
          "librarything": [
            "338356"
          ],
          "isbn_10": [
            "0120644819"
          ],
          "openlibrary": [
            "OL21119585M"
          ]
        },
        "classifications": {
          "lc_classifications": [
            ""
          ]
        },
        "publishers": [
          {
            "name": "AP Professional"
          }
        ],
        "publish_places": [
          {
            "name": "Boston"
          },
          {
            "name": "London"
          }
        ],
        "publish_date": "1991",
        "subjects": [
          {
            "name": "Computer graphics",
            "url": "https://openlibrary.org/subjects/computer_graphics"
          }
        ],
        "ebooks": [
          {
            "preview_url": "https://archive.org/details/graphicsgemsii0000unse_d7q2",
            "availability": "borrow",
            "formats": {
            },
            "borrow_url": "https://openlibrary.org/books/OL21119585M/Graphics_gems_II/borrow",
            "checkedout": true
          }
        ],
        "cover": {
          "small": "https://covers.openlibrary.org/b/id/13821398-S.jpg",
          "medium": "https://covers.openlibrary.org/b/id/13821398-M.jpg",
          "large": "https://covers.openlibrary.org/b/id/13821398-L.jpg"
        }
      },
      "details": {
        "bib_key": "isbn:9780120644810",
        "info_url": "https://openlibrary.org/books/OL21119585M/Graphics_gems_II",
        "preview": "borrow",
        "preview_url": "https://archive.org/details/graphicsgemsii0000unse_d7q2",
        "thumbnail_url": "https://covers.openlibrary.org/b/id/13821398-S.jpg",
        "details": {
          "publishers": [
            "AP Professional"
          ],
          "identifiers": {
            "goodreads": [
              "3994749"
            ],
            "librarything": [
              "338356"
            ]
          },
          "series": [
            "Graphics gems series"
          ],
          "lc_classifications": [
            ""
          ],
          "key": "/books/OL21119585M",
          "publish_places": [
            "Boston",
            "London"
          ],
          "contributions": [
            "Arvo, James, 1956-"
          ],
          "subjects": [
            "Computer graphics."
          ],
          "pagination": "xxxii,643p.,(4)p. of plates :",
          "source_records": [
            "marc:talis_openlibrary_contribution/talis-openlibrary-contribution.mrc:469282395:595",
            "bwb:9780120644810",
            "promise:bwb_daily_pallets_2022-12-15",
            "ia:graphicsgemsii0000unse_d7q2"
          ],
          "title": "Graphics gems II",
          "number_of_pages": 643,
          "languages": [
            {
              "key": "/languages/eng"
            }
          ],
          "isbn_10": [
            "0120644819"
          ],
          "publish_date": "1991",
          "publish_country": "xxu",
          "by_statement": "edited by James Arvo.",
          "works": [
            {
              "key": "/works/OL19309523W"
            }
          ],
          "type": {
            "key": "/type/edition"
          },
          "local_id": [
            "urn:bwbsku:T2-DWT-532"
          ],
          "covers": [
            13821398
          ],
          "ocaid": "graphicsgemsii0000unse_d7q2",
          "latest_revision": 8,
          "revision": 8,
          "created": {
            "type": "/type/datetime",
            "value": "2008-10-31T22:57:16.118970"
          },
          "last_modified": {
            "type": "/type/datetime",
            "value": "2023-04-04T04:33:28.449353"
          }
        }
      }
    }
  },
  "items": [

  ]
}
```

**Comment by ronaldtse** [comment]
Perhaps the https://github.com/jayfajardo/openlibrary gem doesn't use this new API, if that's the case we have to implement a new one.

**Comment by andrew2net** [comment]
> Perhaps the https://github.com/jayfajardo/openlibrary gem doesn't use this new API, if that's the case we have to implement a new one.

I didn't managed to run the gem with Ruby 3.2.
Anyway we need to transform source data to our data model, so we can just use an HTTP request and get data in JSON format. I don't think we need a separate gem for that.

**Comment by andrew2net** [comment]
Done

**Comment by ReesePlews** [comment]
@andrew2net thank you very much for implementing this enhancement!

**Comment by strogonoff** [comment]
@andrew2net any reference as to which version this will be released in?

**Comment by andrew2net** [comment]
> @andrew2net any reference as to which version this will be released in?

@strogonoff it was released and available in relaton and relaton-cli since v 1.17.2

**Comment by ronaldtse** [comment]
@andrew2net can you also update relaton.org for ISBN? Thanks.
